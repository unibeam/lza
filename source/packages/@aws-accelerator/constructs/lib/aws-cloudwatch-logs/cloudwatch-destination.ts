/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

/**
 * Construction properties for CloudWatch Destination for Kinesis Stream.
 */
export interface CloudWatchDestinationProps {
  /**
   *
   * KMS key to encrypt the Kinesis Data Stream
   */
  kinesisKmsKey: cdk.aws_kms.IKey;
  /**
   *
   * Kinesis Data Stream for CloudWatch logs
   */
  kinesisStream: cdk.aws_kinesis.IStream;
  /**
   * Organization ID to restrict the usage within specific org
   */
  organizationId?: string;
  /**
   * Partition to determine the IAM condition.
   */
  partition: string;
  /**
   * Account IDs for the IAM condition.
   */
  accountIds?: string[];
  /**
   * Accelerator Prefix defaults to 'AWSAccelerator'.
   */
  acceleratorPrefix: string;
  /**
   * Use existing IAM roles for deployment.
   */
  useExistingRoles: boolean;
}
/**
 * Class to configure CloudWatch Destination on logs receiving account
 */
export class CloudWatchDestination extends Construct {
  constructor(scope: Construct, id: string, props: CloudWatchDestinationProps) {
    super(scope, id);

    let principalOrgIdCondition: object | undefined = undefined;
    let accountPrincipals: object | cdk.aws_iam.IPrincipal;

    if (props.partition === 'aws-cn' || !props.organizationId) {
      // Only principal block with list of account id is supported.
      accountPrincipals = {
        AWS: props.accountIds,
      };
    } else {
      principalOrgIdCondition = {
        StringEquals: {
          'aws:PrincipalOrgID': props.organizationId,
        },
      };
      accountPrincipals = new cdk.aws_iam.AnyPrincipal();
    }

    const logDestinationPolicy = JSON.stringify({
      Version: '2012-10-17',
      Statement: [
        {
          Effect: 'Allow',
          Principal: accountPrincipals,
          Action: 'logs:PutSubscriptionFilter',
          Resource: `arn:${cdk.Stack.of(this).partition}:logs:${cdk.Stack.of(this).region}:${
            cdk.Stack.of(this).account
          }:destination:${props.acceleratorPrefix}CloudWatchToS3`,
          Condition: principalOrgIdCondition,
        },
      ],
    });

    new cdk.aws_logs.CfnDestination(this, 'Resource', {
      roleArn: this.createKinesisRole(
        props.kinesisStream.streamArn,
        props.kinesisKmsKey.keyArn,
        props.useExistingRoles,
        props.acceleratorPrefix,
      ),
      targetArn: props.kinesisStream.streamArn,
      destinationName: `${props.acceleratorPrefix}CloudWatchToS3`,
      destinationPolicy: logDestinationPolicy,
    });
  }

  private createKinesisRole(
    kinesisStreamArn: string,
    kinesisKeyArn: string,
    useExistingRoles: boolean,
    acceleratorPrefix: string,
  ) {
    if (useExistingRoles) {
      return `arn:${cdk.Stack.of(this).partition}:iam::${
        cdk.Stack.of(this).account
      }:role/${acceleratorPrefix}LogReplicationRole-${cdk.Stack.of(this).region}`;
    }
    //Create policy for access to Kinesis stream
    const kinesisStreamAccess = new cdk.aws_iam.PolicyDocument({
      statements: [
        new cdk.aws_iam.PolicyStatement({
          actions: ['kinesis:ListShards', 'kinesis:PutRecord', 'kinesis:PutRecords'],
          resources: [kinesisStreamArn],
        }),
      ],
    });
    const kmsKeyAccess = new cdk.aws_iam.PolicyDocument({
      statements: [
        new cdk.aws_iam.PolicyStatement({
          actions: [
            'kms:Decrypt',
            'kms:Encrypt',
            'kms:GenerateDataKey',
            'kms:ReEncryptTo',
            'kms:GenerateDataKeyWithoutPlaintext',
            'kms:GenerateDataKeyPairWithoutPlaintext',
            'kms:GenerateDataKeyPair',
            'kms:ReEncryptFrom',
          ],
          resources: [kinesisKeyArn],
        }),
      ],
    });
    // Create a role for CloudWatch Logs destination
    const logsKinesisRole = new cdk.aws_iam.Role(this, 'LogsKinesisRole', {
      assumedBy: new cdk.aws_iam.ServicePrincipal(`logs.${cdk.Stack.of(this).region}.amazonaws.com`),
      // this needs to be inline to ensure role is created with proper access
      // if not, CloudWatch destination creation fails with no permission to access Kinesis or KMS (generateDataKey access error)
      inlinePolicies: {
        KinesisAccess: kinesisStreamAccess,
        KmsAccess: kmsKeyAccess,
      },
    });
    return logsKinesisRole.roleArn;
  }
}
