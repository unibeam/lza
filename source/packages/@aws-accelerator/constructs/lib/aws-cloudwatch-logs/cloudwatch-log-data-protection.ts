/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { LzaCustomResource } from '../lza-custom-resource';
import * as path from 'path';

/**
 * Construction properties for CloudWatch CloudWatchLogDataProtection.
 */
export interface CloudWatchLogDataProtectionProps {
  /**
   *
   * Central logs bucket name
   */
  centralLogBucketName: string;
  /**
   *
   * Data protection identifier names
   */
  identifierNames: string[];
  /**
   * Indicates whether existing CloudWatch Log data protection policy configuration can be overwritten.
   */
  overrideExisting: boolean;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly customResourceLambdaEnvironmentEncryptionKmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly customResourceLambdaCloudWatchLogKmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly customResourceLambdaLogRetentionInDays: number;
}
/**
 * Class to configure CloudWatch log data protection
 */
export class CloudWatchLogDataProtection extends Construct {
  constructor(scope: Construct, id: string, props: CloudWatchLogDataProtectionProps) {
    super(scope, id);

    const resourceName = 'CloudWatchDataProtection';

    new LzaCustomResource(this, resourceName, {
      resource: {
        name: resourceName,
        parentId: id,
        properties: [
          {
            centralLogBucketName: props.centralLogBucketName,
            identifierNames: props.identifierNames,
            partition: cdk.Stack.of(this).partition,
            overrideExisting: props.overrideExisting,
          },
        ],
      },
      lambda: {
        assetPath: path.join(__dirname, 'put-account-policy/dist'),
        environmentEncryptionKmsKey: props.customResourceLambdaEnvironmentEncryptionKmsKey,
        cloudWatchLogKmsKey: props.customResourceLambdaCloudWatchLogKmsKey,
        cloudWatchLogRetentionInDays: props.customResourceLambdaLogRetentionInDays,
        timeOut: cdk.Duration.minutes(15),
        roleInitialPolicy: [
          new cdk.aws_iam.PolicyStatement({
            sid: 'CloudWatchAccess',
            effect: cdk.aws_iam.Effect.ALLOW,
            actions: [
              'logs:DeleteAccountPolicy',
              'logs:DescribeAccountPolicies',
              'logs:PutAccountPolicy',
              'logs:PutDataProtectionPolicy',
              'logs:DeleteDataProtectionPolicy',
              'logs:CreateLogDelivery',
            ],
            resources: ['*'],
          }),
          new cdk.aws_iam.PolicyStatement({
            sid: 'S3Access',
            effect: cdk.aws_iam.Effect.ALLOW,
            actions: ['s3:GetBucketPolicy', 's3:PutBucketPolicy'],
            resources: [`arn:${cdk.Stack.of(this).partition}:s3:::${props.centralLogBucketName}`],
          }),
        ],
      },
    });
  }
}
