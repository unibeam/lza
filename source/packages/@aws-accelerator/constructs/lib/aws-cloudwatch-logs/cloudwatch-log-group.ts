/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as cdk from 'aws-cdk-lib';
import * as path from 'path';
import { pascalCase } from 'change-case';
import { Construct } from 'constructs';

/**
 * Construction properties for CloudWatch LogGroups.
 */

export type LogGroupRetention =
  | 1
  | 3
  | 5
  | 7
  | 14
  | 30
  | 60
  | 90
  | 120
  | 150
  | 180
  | 365
  | 400
  | 545
  | 731
  | 1096
  | 1827
  | 2192
  | 2557
  | 2922
  | 3288
  | 3653;

export interface CloudWatchLogGroupsProps {
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly customLambdaLogKmsKey?: cdk.aws_kms.IKey;
  /**
   * How long, in days, the log contents for the Lambda function
   * will be retained.
   */
  readonly customLambdaLogRetention: number;
  /**
   * How long, in days, the log contents will be retained.
   *
   * To retain all logs, set this value to undefined.
   *
   */
  readonly logRetentionInDays: LogGroupRetention | number;
  /**
   * KMS Key Arn to encrypt CloudWatch Logs Group at rest.
   */
  readonly keyArn?: string;
  /**
   *
   * Name of the CloudWatch Logs Group
   */
  readonly logGroupName?: string;
  /**
   * For cross-account log groups, the owning account ID
   */
  readonly owningAccountId?: string;
  /**
   * For cross-region log groups, the owning region
   */
  readonly owningRegion?: string;
  /**
   * For cross-account log groups, the IAM role name to assume
   */
  readonly roleName?: string;
  /**
   *
   * Determine termination policy on CloudWatch Logs Group
   */
  readonly terminationProtected?: boolean;
}

interface ILogGroup {
  /**
   * The name of the log group
   */
  readonly logGroupName: string;
  /**
   * The ARN of the log group
   */
  readonly logGroupArn: string;
}

/**
 * Class to configure CloudWatch Log Groups
 */
export class CloudWatchLogGroups extends cdk.Resource implements ILogGroup {
  public readonly logGroupName: string;
  public readonly logGroupArn: string;

  constructor(scope: Construct, id: string, props: CloudWatchLogGroupsProps) {
    super(scope, id);
    const CLOUD_WATCH_LOG_GROUPS = 'Custom::CreateLogGroup';
    this.logGroupName = props.logGroupName ?? cdk.Names.uniqueResourceName(this, { separator: '-' });

    //
    // Function definition for the custom resource
    //
    const policyStatements = props.owningAccountId
      ? [
          {
            Effect: 'Allow',
            Action: ['sts:AssumeRole'],
            Resource: `arn:${this.stack.partition}:iam::*:role/${props.roleName}`,
          },
        ]
      : [
          {
            Effect: 'Allow',
            Action: ['logs:CreateLogGroup', 'logs:DeleteLogGroup', 'logs:PutRetentionPolicy'],
            Resource: `arn:${this.stack.partition}:logs:${props.owningRegion ?? this.stack.region}:${
              this.stack.account
            }:log-group:${this.logGroupName}:*`,
          },
          {
            Effect: 'Allow',
            Action: ['kms:DescribeKey', 'kms:ListKeys', 'logs:AssociateKmsKey', 'logs:DescribeLogGroups'],
            Resource: '*',
          },
        ];

    const provider = cdk.CustomResourceProvider.getOrCreateProvider(
      this,
      pascalCase(`${this.logGroupName}-${CLOUD_WATCH_LOG_GROUPS}`),
      {
        codeDirectory: path.join(__dirname, 'create-log-groups/dist'),
        runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
        policyStatements,
      },
    );

    const resource = new cdk.CustomResource(this, 'CloudWatchLogsResource', {
      resourceType: CLOUD_WATCH_LOG_GROUPS,
      serviceToken: provider.serviceToken,
      properties: {
        logGroupName: this.logGroupName,
        retention: props.logRetentionInDays,
        keyArn: props.keyArn,
        owningAccountId: props.owningAccountId,
        owningRegion: props.owningRegion,
        roleName: props.roleName,
        terminationProtected: props.terminationProtected,
      },
    });
    this.logGroupArn = resource.getAttString('LogGroupArn');

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.customLambdaLogRetention,
        encryptionKey: props.customLambdaLogKmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);
  }
}
