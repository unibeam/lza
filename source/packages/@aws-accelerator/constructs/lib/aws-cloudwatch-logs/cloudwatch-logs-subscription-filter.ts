/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as path from 'path';

export type cloudwatchExclusionItem = {
  account: string;
  region: string;
  excludeAll?: boolean;
  logGroupNames?: string[];
};

/**
 * Construction properties for CloudWatch Logs Creating account.
 */

export interface CloudWatchLogsCreateProps {
  /**
   *
   * Log Destination Arn to which all the logs will get forwarded to
   */
  readonly logDestinationArn: string;
  /**
   *
   * KMS key to encrypt the Lambda environment variables, when undefined default AWS managed key will be used
   */
  readonly logsKmsKey?: cdk.aws_kms.IKey;
  /**
   *
   * CloudWatch Retention in days from global config
   */
  readonly logsRetentionInDays: string;
  /**
   *
   * Subscription Filter Arn
   */
  readonly subscriptionFilterRoleArn: string;
  /**
   *
   * LogArchive account Id
   */
  readonly logArchiveAccountId: string;
  /**
   * CloudWatch Logs exclusion options
   */
  readonly logExclusionOption?: cloudwatchExclusionItem;
  /**
   * Existing customer defined log subscription destination arn, which accelerator needs to remove before configuring solution defined subscription destination.
   *
   */
  readonly replaceLogDestinationArn?: string;
  /**
   * Accelerator Prefix defaults to 'AWSAccelerator'.
   */
  readonly acceleratorPrefix: string;
  /**
   * Use existing IAM roles for deployment.
   */
  readonly useExistingRoles: boolean;
}

/**
 * Class to configure CloudWatch Destination on logs receiving account
 */
export class CloudWatchLogsSubscriptionFilter extends Construct {
  readonly id: string;
  constructor(scope: Construct, id: string, props: CloudWatchLogsCreateProps) {
    super(scope, id);
    let acceleratorLogSubscriptionRoleArn: string;
    if (props.useExistingRoles) {
      acceleratorLogSubscriptionRoleArn = `arn:${cdk.Stack.of(this).partition}:iam::${
        cdk.Stack.of(this).account
      }:role/${props.acceleratorPrefix}LogReplicationRole-${cdk.Stack.of(this).region}`;
    } else {
      acceleratorLogSubscriptionRoleArn = props.subscriptionFilterRoleArn;
    }

    const UPDATE_SUBSCRIPTION_FILTER = 'Custom::UpdateSubscriptionFilter';

    //
    // Function definition for the custom resource
    //
    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, UPDATE_SUBSCRIPTION_FILTER, {
      codeDirectory: path.join(__dirname, 'update-subscription-filter/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      policyStatements: [
        // Required when global-config.yaml::logging::cloudwatchLogs::encryption is configured
        {
          Sid: 'AllowEncryption',
          Effect: 'Allow',
          Action: ['kms:Encrypt', 'kms:GenerateDataKey'],
          Resource: [
            `arn:${cdk.Stack.of(this).partition}:kms:${cdk.Stack.of(this).region}:${cdk.Stack.of(this).account}:key:*`,
          ],
        },

        {
          Effect: 'Allow',
          Action: [
            'logs:PutRetentionPolicy',
            'logs:AssociateKmsKey',
            'logs:DescribeLogGroups',
            'logs:DescribeSubscriptionFilters',
          ],
          Resource: [
            `arn:${cdk.Stack.of(this).partition}:logs:${cdk.Stack.of(this).region}:${
              cdk.Stack.of(this).account
            }:log-group:*`,
          ],
        },

        {
          Effect: 'Allow',
          Action: ['logs:PutSubscriptionFilter', 'logs:DeleteSubscriptionFilter'],
          Resource: [
            `arn:${cdk.Stack.of(this).partition}:logs:${cdk.Stack.of(this).region}:${
              cdk.Stack.of(this).account
            }:log-group:*`,
            `arn:${cdk.Stack.of(this).partition}:logs:${cdk.Stack.of(this).region}:${
              props.logArchiveAccountId
            }:destination:*`,
          ],
        },
      ],
    });
    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: UPDATE_SUBSCRIPTION_FILTER,
      serviceToken: provider.serviceToken,
      properties: {
        acceleratorLogRetentionInDays: props.logsRetentionInDays,
        acceleratorCreatedLogDestinationArn: props.logDestinationArn,
        acceleratorLogSubscriptionRoleArn,
        acceleratorLogKmsKeyArn: props.logsKmsKey ? props.logsKmsKey.keyArn : undefined,
        logExclusionOption: props.logExclusionOption
          ? JSON.stringify(props.logExclusionOption)
          : props.logExclusionOption,
        replaceLogDestinationArn: props.replaceLogDestinationArn,
      },
    });

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: parseInt(props.logsRetentionInDays),
        encryptionKey: props.logsKmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);

    this.id = resource.ref;
  }
}
