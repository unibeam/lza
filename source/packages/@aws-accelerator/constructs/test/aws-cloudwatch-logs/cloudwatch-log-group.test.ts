/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { CloudWatchLogGroups } from '../../lib/aws-cloudwatch-logs/cloudwatch-log-group';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(CloudWatchLogGroups): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

new CloudWatchLogGroups(stack, 'CloudWatchLogGroups', {
  logGroupName: '/App/Test1',
  logRetentionInDays: 30,
  keyArn: 'arn:aws:kms:us-east-1:111111111111:key/121ac3b6-8d53-4d8a-a05c-7a0012249950',
  customLambdaLogKmsKey: new cdk.aws_kms.Key(stack, 'CustomKey', {}),
  customLambdaLogRetention: 30,
});

new CloudWatchLogGroups(stack, 'CloudWatchLogGroupsNoKey', {
  logGroupName: '/App/Test2',
  logRetentionInDays: 180,
  customLambdaLogKmsKey: new cdk.aws_kms.Key(stack, 'CloudWatchKey', {}),
  customLambdaLogRetention: 30,
});

/**
 * CloudWatchLogGroups construct test
 */
describe('CloudWatchLogGroups', () => {
  snapShotTest(testNamePrefix, stack);
});
