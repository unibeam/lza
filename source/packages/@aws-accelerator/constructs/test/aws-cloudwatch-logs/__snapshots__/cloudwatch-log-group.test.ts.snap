// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudWatchLogGroups Construct(CloudWatchLogGroups):  Snapshot Test 1`] = `
{
  "Resources": {
    "AppTest1CustomCreateLogGroupCustomResourceProviderHandler4FFCDD76": {
      "DependsOn": [
        "AppTest1CustomCreateLogGroupCustomResourceProviderRole40E7A5EC",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "AppTest1CustomCreateLogGroupCustomResourceProviderRole40E7A5EC",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AppTest1CustomCreateLogGroupCustomResourceProviderLogGroupF4CE828F": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "CustomKey1E6D0D07",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "AppTest1CustomCreateLogGroupCustomResourceProviderHandler4FFCDD76",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "AppTest1CustomCreateLogGroupCustomResourceProviderRole40E7A5EC": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "logs:CreateLogGroup",
                    "logs:DeleteLogGroup",
                    "logs:PutRetentionPolicy",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/App/Test1:*",
                      ],
                    ],
                  },
                },
                {
                  "Action": [
                    "kms:DescribeKey",
                    "kms:ListKeys",
                    "logs:AssociateKmsKey",
                    "logs:DescribeLogGroups",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AppTest2CustomCreateLogGroupCustomResourceProviderHandler03798759": {
      "DependsOn": [
        "AppTest2CustomCreateLogGroupCustomResourceProviderRole24FD6060",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "AppTest2CustomCreateLogGroupCustomResourceProviderRole24FD6060",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AppTest2CustomCreateLogGroupCustomResourceProviderLogGroup7FB49646": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "CloudWatchKey9B181885",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "AppTest2CustomCreateLogGroupCustomResourceProviderHandler03798759",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "AppTest2CustomCreateLogGroupCustomResourceProviderRole24FD6060": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "logs:CreateLogGroup",
                    "logs:DeleteLogGroup",
                    "logs:PutRetentionPolicy",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/App/Test2:*",
                      ],
                    ],
                  },
                },
                {
                  "Action": [
                    "kms:DescribeKey",
                    "kms:ListKeys",
                    "logs:AssociateKmsKey",
                    "logs:DescribeLogGroups",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CloudWatchKey9B181885": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "CloudWatchLogGroupsCloudWatchLogsResourceB533E7A1": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "AppTest1CustomCreateLogGroupCustomResourceProviderLogGroupF4CE828F",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "AppTest1CustomCreateLogGroupCustomResourceProviderHandler4FFCDD76",
            "Arn",
          ],
        },
        "keyArn": "REPLACED-UUID",
        "logGroupName": "/App/Test1",
        "retention": 30,
      },
      "Type": "Custom::CreateLogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CloudWatchLogGroupsNoKeyCloudWatchLogsResource3EFE207D": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "AppTest2CustomCreateLogGroupCustomResourceProviderLogGroup7FB49646",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "AppTest2CustomCreateLogGroupCustomResourceProviderHandler03798759",
            "Arn",
          ],
        },
        "logGroupName": "/App/Test2",
        "retention": 180,
      },
      "Type": "Custom::CreateLogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomKey1E6D0D07": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
